### PR TITLE
[Merged by Bors] - add get_history function to Diagnostic

### LINKED ISSUE #2771: Access Diagnostic::history field
## What problem does this solve or what need does it fill?

I would like to access the history data of a diagnostic for drawing a graph using egui.

## What solution would you like?

The history field should be made public. Or a function to access the underlying history data.

## What alternative(s) have you considered?

Currently I'm storing the Diagnostic value in separate resource but this would be unnecessary if i could access the history stored with the diagnostic.

### PR DESCRIPTION
**Pull request by bytebuddha**
# Objective

- Allow access to diagnostic history value.
- Fixes #2771.

## Solution

- Add Diagnostic::get_history function.

### CODE REVIEW

File: crates/bevy_diagnostic/src/diagnostic.rs [hunk 1]
```diff
@@ -119,6 +119,10 @@ impl Diagnostic {
     pub fn get_max_history_length(&self) -> usize {
         self.max_history_length
     }
+
+    pub fn get_history(&self) -> Vec<f64> {
```
Comment (alice-i-cecile) [inline crates/bevy_diagnostic/src/diagnostic.rs:123]:
```suggestion
    pub fn history(&self) -> Vec<f64> {
```

Elsewhere in the code base we tend to prefer this naming convention.

**Comment by alice-i-cecile** [comment]
Two quick questions about the return type:

1. Why do we want to collect this into a vec, rather rather returning an iterator?
2. Are we okay to discard the temporal data here? I suspect it's going to be essential for more complex use cases.

I think that two separate methods make sense here; one for just the values, called `.values`, and a second one with the temporal data called `.history` that just returns the `history` field.

**Comment by bytebuddha** [comment]
Although i haven't had a use for it yet, I think your completely right. Just to be clear you are talking about two functions signatures like this right?

```rust
pub fn values(&self) -> impl Iterator<Item=&f64>;
pub fn history(&self) -> impl Iterator<Item=&DiagnosticMeasurement>
```

**Comment by cart** [comment]
Yup I agree that we should be using iterators here instead of allocating. `values` and `history` are both "history", one just includes more information. I think these names might be a bit less ambiguous:

```rust
pub fn values(&self) -> impl Iterator<Item=&f64>;
pub fn measurements(&self) -> impl Iterator<Item=&DiagnosticMeasurement>
```

Using "measurements" also has the benefit of having parity with `add_measurement`. However they are both still "measurements". Maybe instead we should just expose a single  `pub fn measurements(&self) -> impl Iterator<Item=&DiagnosticMeasurement>` and let users access the values via the iterator.

**Comment by alice-i-cecile** [comment]
> Just to be clear you are talking about two functions signatures like this right?

Correct :) I also prefer Cart's `measurements` naming suggestion. I could go either way on the one vs two methods. I think one method gets a slight preference for me because I like the type-safety of `&DiagnosticMeasurement` and it's trivial to extract the underlying data if that's all you want.

**Comment by bytebuddha** [comment]
Sorry for taking a while to respond, I agree with all the points made and have made the changes. I believe this should be ready for a final look then

**Review by alice-i-cecile — APPROVED** [review]

**Comment by cart** [comment]
Haha just realized this will fail. @bytebuddha can you run `cargo fmt --all` and commit the changes?

**Pull request closed without merging**